### PR TITLE
EHN: GAT parallel across chunks for predict & score

### PR DESCRIPTION
**Pull request by kingjr**
@dengemann @jona-sassenhagen @cmoutard

The previous implementation was parallelized across time samples, which was highly suboptimal when dealing with simple classifiers (e.g. based on a low number of features, like with eeg) and a lot of time points, because of constant jumps across CPU threads.

With the current implementation, the `predict` and `score` parallelization is made with chunked data. This is analogous to the method used for `fit`, but the chunked data is across time (not estimators), which is slightly more difficult to follow, but allows dispatching `X` across threads, as opposed to copy it in each thread.

The cpu and memory gain is pretty high for large datasets. (@agramfort No, I haven't done the profiling, but I see it by eye ;))

I apologize in advance: the syntax remains quite hard to follow, most of the time because I made the stupid decision early on to accept irregular training and testing times (which can be useful, but not that often). There may be room for improvement and clarification (notably by using a single function to chunk and parallelize the data), but this will be for a later PR.

(Also: I've had problems with the git history, some old commits were omitted; I'm not sure why, I tried rebasing a few time, keeping silent features, checked the history,  etc; in vain. I think it's because i) I started this branch a while ago and/or ii) I continued from https://github.com/mne-tools/mne-python/pull/2398)

Let me know what you think.

### CODE REVIEW

File: mne/decoding/time_gen.py [hunk 1]
```diff
@@ -614,7 +651,7 @@ def _predict(X, estimators):
             predict_size = _y_pred.shape[1] if _y_pred.ndim > 1 else 1
             y_pred = np.ones((n_epochs, predict_size, n_clf))
         if predict_size == 1:
-            y_pred[:, 0, fold] = _y_pred
+            y_pred[:, 0, fold] = np.squeeze(_y_pred)
```
Comment (kingjr) [inline mne/decoding/time_gen.py]: this will need the rebase from https://github.com/mne-tools/mne-python/pull/2398

**Comment by jona-sassenhagen** [comment]
Nice - is this read for merge? I've taken a quick look and didn't spot anything.

It seems #2398 is also ready but for two lines?

**Comment by kingjr** [comment]
> Nice - is this read for merge? I've taken a quick look and didn't spot anything.

I think so. Have you tried it?

> It seems #2398 is also ready but for two lines?

Yes, I'm behind my own agenda..

**Comment by agramfort** [comment]
please rebase.

File: mne/decoding/time_gen.py [hunk 2]
```diff
@@ -311,18 +330,32 @@ def score(self, epochs=None, y=None):
...
-        return scores
+def _predict_slices(X, estimators, cv, slices, predict_mode):
+    """Aux function of GeneralizationAcrossTime that loops across chunks of
+    testing slices.
+    """
+    out = list()
+    for this_estimator, this_slice in zip(estimators, slices):
+        out.append(_predict_time_loop(X, this_estimator, cv, this_slice,
+                                      predict_mode))
+    return out
```
Comment (jona-sassenhagen) [inline mne/decoding/time_gen.py]: Is this considered cleaner than
```
return [_predict_time_loop(X, this_estimator, cv, this_slice, predict_mode)
        for this_estimator, this_slice in zip(estimators, slices)]
````
?

Comment (agramfort) [inline mne/decoding/time_gen.py]: I would use a list comprehension here.

**Comment by agramfort** [comment]
Thanks @kingjr 

Sorry for being slow to review

File: mne/decoding/time_gen.py (hunk 2 above)
Comment (kingjr) [inline mne/decoding/time_gen.py]: I had missed that, will try to keep it in mind in the next PR

**Comment by kingjr** [comment]
no worries & thank you ;)
